[Merged by Bors] - feat(linear_algebra/quadratic_form): more constructions for quadratic forms

Vierkantor opened the pull request:
Define multiplication of two linear forms to give a quadratic form and addition of quadratic forms. With these definitions, we can write a generic binary quadratic form as `a • proj R₁ 0 0 + b • proj R₁ 0 1 + c • proj R₁ 1 1 : quadratic_form R₁ (fin 2 → R₁)`.

In order to prove the linearity conditions on the constructions, there are new `simp` lemmas `polar_add_left`, `polar_smul_left`, `polar_add_right` and `polar_smul_right` copying from the corresponding fields of the `quadratic_form` structure, that use `⇑ Q` instead of `Q.to_fun`. The original field names have a `'` appended to avoid name clashes.

Vierkantor committed:
feat(linear_algebra/quadratic_form): more constructions for quadratic forms

Define multiplication of two linear forms to give a quadratic form and addition
of quadratic forms. With these definitions, we can write a generic binary
quadratic form as
`a • proj R₁ 0 0 + b • proj R₁ 0 1 + c • proj R₁ 1 1 : quadratic_form R₁ (fin 2 → R₁)`.

In order to prove the linearity conditions on the constructions, there are new
`simp` lemmas `polar_add_left`, `polar_smul_left`, `polar_add_right` and
`polar_smul_right` copying from the corresponding fields of the
`quadratic_form` structure, that use `⇑ Q` instead of `Q.to_fun`.
The original field names have a `'` appended to avoid name clashes.

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -52,26 +52,36 @@ quadratic form, homogeneous polynomial, quadratic polynomial
-(polar_smul_left : ∀ (a : R) (x y : M), polar to_fun (a • x) y = a • polar to_fun x y)
-(polar_add_right : ∀ (x y y' : M), polar to_fun x (y + y') = polar to_fun x y + polar to_fun x y')
-(polar_smul_right : ∀ (a : R) (x y : M), polar to_fun x (a • y) = a • polar to_fun x y)
+(polar_add_left' : ∀ (x x' y : M), polar to_fun (x + x') y = polar to_fun x y + polar to_fun x' y)
```
I think another option would be to make the structure fields protected with the new `@[protect_proj]` attribute on this structure. And then add the simp lemmas to the root namespace. (I don't expect future name clashes with `polar_add_left` in the root namespace.)
But the current approach is a bit more conservative, which is also fine.

Vierkantor replied:
That's also a nice solution. I think I prefer to keep everything in the `quadratic_form` namespace, just like in the other `..._form` definitions. Do we have guidelines on how to structure namespaces?

jcommelin replied:
Not really... I think we're still learning while doing. (Apart from structuring things in such a way that dot notation works nicely.)
Now that I think of it, dot notation is another reason to keep it this way.

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -139,21 +199,80 @@ variables {N : Type v} [add_comm_group N] [module R N]
+/-- The product of linear forms is a quadratic form. -/
+def lin_mul_lin (f g : M →ₗ[R₁] R₁) : quadratic_form R₁ M :=
+mk_left (f * g)
+  (λ a x, by { simp, ring } )
```
```suggestion
  (λ a x, by { simp, ring })
```

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -139,21 +199,80 @@ variables {N : Type v} [add_comm_group N] [module R N]
+  (λ a x y, by { simp [polar], ring })
+
+@[simp]
+lemma lin_mul_lin_map (f g : M →ₗ[R₁] R₁) (x) : lin_mul_lin f g x = f x * g x := rfl
```
```suggestion
lemma lin_mul_lin_apply (f g : M →ₗ[R₁] R₁) (x) : lin_mul_lin f g x = f x * g x := rfl
```

Vierkantor replied:
It looks like some other definitions in this file also use `_map`. Should I change those as well?

jcommelin replied:
Yes please. I think the emerging rule is `_apply` is for the (simp) lemma that says what happens if you apply a bundled morphism to an input. But `map_{mul,add,id,zero,one,...}` is saying that a morphism preserves some "structure".

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -247,6 +375,17 @@ lemma smul_pos_def_of_nonzero {K : Type u} [linear_ordered_field K] [module K M]
 
+variables {n : Type*}
+
+lemma add_pos_def (Q Q' : quadratic_form R₂ M) (hQ : pos_def Q) (hQ' : pos_def Q') :
```
You could change this to `pos_def.add`, so that people can use it as `hQ.add hQ'`.

Vierkantor committed:
Apply suggestions

Co-authored-by: Johan Commelin <johan@commelin.net>

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -104,31 +134,61 @@ by { cases Q, cases Q', congr, funext, apply H }
+  ..quadratic_form.has_add,
+  ..quadratic_form.has_zero }
+
+instance : distrib_mul_action R₁ (quadratic_form R₁ M) :=
```
Don't you even get a full blown module structure?

Vierkantor replied:
We should, but I defined the wrong `has_scalar` instance before :P

Vierkantor committed:
Fix the `has_scalar R (quadratic_form R M)` instance

`smul` was defined as a scalar multiplication of the argument, but of course
it should be of the result. Now we can define the `module` instance.

jcommelin commented on src/linear_algebra/quadratic_form.lean:
```diff
@@ -184,12 +188,13 @@ instance : add_comm_monoid (quadratic_form R M) :=
 
-instance : distrib_mul_action R₁ (quadratic_form R₁ M) :=
-{ mul_smul := λ a b Q, ext (λ x, by simp [mul_smul, smul_comm]),
+instance : semimodule R₁ (quadratic_form R₁ M) :=
```
It's actually a module, right? You have a negation on `quadratic_form _ _`.

Vierkantor replied:
You are completely right! I was making the same mistake as for `has_scalar`.

jcommelin replied:
Awesome! Thanks!

Vierkantor committed:
Add the `has_neg` and `module` instances on `quadratic_form`

jcommelin reviewed (approved):
Thanks :tada:

bors merge
